Added logger abstraction with default implementation

serg-temchenko opened the pull request:
- Added logger abstraction
- Added default logger implementation

Related to PR #129 

Work which already done with defining places to log in Agent and in InteractionManager/Interaction currently stashed and will wait that moment when service manager will be integrated and `DefaultLogger` instance will be available to use...

serg-temchenko committed:
feat(logger): added required libraries for the logging integration

serg-temchenko committed:
feat(logger): added abstraction of the logger

serg-temchenko committed:
feat(logger): defined logger types

serg-temchenko committed:
feat(logger): defined default logger implementation

serg-temchenko committed:
feat(logger): added barrel export

mnzaki commented on package.json:
```diff
@@ -34,6 +35,7 @@
     "@typescript-eslint/eslint-plugin": "^2.20.0",
     "@typescript-eslint/parser": "^1.5.0",
     "@zerollup/ts-transform-paths": "^1.7.17",
+    "app-root-path": "^3.0.0",
```
what's this for?

serg-temchenko replied:
To define path to the default place to save logs [here](https://github.com/jolocom/jolocom-sdk/pull/134/files#diff-cf09c68e036672514f763df6d940fd8635a7f66ef1fe5e29b6913f6773132a54R49)

mnzaki commented on package.json:
```diff
@@ -45,11 +47,12 @@
     "prettier": "^1.18.2",
     "regenerator-runtime": "^0.13.5",
-    "sqlite3": "^4",
+    "sqlite3": "^5.0.2",
```
why this version bump in this PR?

serg-temchenko replied:
For some reason yarn bumped version on adding other dependency, will remove this change

serg-temchenko replied:
Removed from PR

mnzaki commented on package.json:
```diff
@@ -45,11 +47,12 @@
     "fs-extra": "^8.1.0",
     "jest": "^26.4.2",
     "jolocom-lib": "^5.2.1",
+    "log4js": "^6.3.0",
```
have you compared this against other options for logging libraries? Why did you choose log4js in specific?

serg-temchenko replied:
yes, I compered with others (with [winston](https://www.npmjs.com/package/winston) specifically) and [log4js](https://www.npmjs.com/package/log4js) more lightweight (~x3 times), has standardized log structure out of the box, easier configurable and faster (for example [those](https://www.loggly.com/blog/a-benchmark-of-five-node-js-logging-libraries/) results)

mnzaki reviewed (changes requested):
Left some inline comments. Furthermore:

- please squash redundant commits, such as (but not limited to) `feat(logger): fixed logger arguments types`
- for the types/interfaces/enums I think they should be combined in `types.ts` instead of splitting many tiny definitions across many files
- finally, I think the enum method for defining channels is kinda limiting. For example, when a plugin wants to log something, we should be able to see what plugin it is, or better yet, which file is logging! In other words, this should be easily extendible by different subparts of the SDK and different plugins and even user-defined plugins of course!


Please actually compare other libs and logging methods, as I think `log4js` seems a bit ummm old/hacky. A quick look at it shows that it was started for the browser and was ported to node. I have personally used `winston` before, and I think `pino` was recommended as well, please compare them first

serg-temchenko commented:
> Left some inline comments. Furthermore:
> 
> * please squash redundant commits, such as (but not limited to) `feat(logger): fixed logger arguments types`
> * for the types/interfaces/enums I think they should be combined in `types.ts` instead of splitting many tiny definitions across many files
> * finally, I think the enum method for defining channels is kinda limiting. For example, when a plugin wants to log something, we should be able to see what plugin it is, or better yet, which file is logging! In other words, this should be easily extendible by different subparts of the SDK and different plugins and even user-defined plugins of course!
> 
> Please actually compare other libs and logging methods, as I think `log4js` seems a bit ummm old/hacky. A quick look at it shows that it was started for the browser and was ported to node. I have personally used `winston` before, and I think `pino` was recommended as well, please compare them first

About 'squash' - IMO it's ok to have full history of changes, but ok, I squashed those commits...

About ' types/interfaces/enums' in one file - not just the logic must be decoupled and encapsulated, but the structure as well, I don't see some profit to put everything in one file (like classes/types/interfaces etc.) it's definitely the 'bad smell' which brings confusions and bad structure in overall - but ok, I put everything in one file...

About 'enum method for defining channels' - this is the definition of predefined configuration where we already know about the channels and the type of logs to store, additionally I added ['customConfig'](https://github.com/jolocom/jolocom-sdk/pull/134/commits/616036fcb6c33235901d94504b182a40a927b385#diff-ea82e25c2e33c37b4ff1ec1a6ee3c8684e0c76a2e78a1d3c82096546e6c0ae80R53) property to be configured from the client app so when there will be defined 'service management system' all we need is just iterate over all 'categories' and create instances of logger channels objects and request it from any part of the app/plugin and use... 

About 'which file is logging' - I assume the channel definition in log record is more than enough to understand which plugin/part of the code relates to this log.... in case if we logging errors -  stack trace could be taken from catched error.

About 'comparing libs' - not sure that I understand argue about 'old/hacky' (winston actually older than log4js). IMO defining/discussing/suggesting of the lib to use must be done on the planing stage but not when implementation is done if there is some preference... but about comparing libs I answered [here](https://github.com/jolocom/jolocom-sdk/pull/134#discussion_r720320527), please take a look.
